gbmi/exp_max_of_n/train.py

```diff
@@ -0,0 +1,223 @@
+
+
+MAX_OF_2_CONFIG = Config(
+    experiment=MaxOfN(
+        force_adjacent=True,
+    ),
+    n_ctx=2,
+)
+MAX_OF_10_CONFIG = Config(
+    experiment=MaxOfN(
+        n_train_samples=None,
+        n_test_samples=1024,
+    ),
+    n_ctx=10,
+)
+
+
+class MaxOfNTrainingWrapper(TrainingWrapper[MaxOfN]):
+    def __init__(self, config: Config[MaxOfN], model: HookedTransformer):
+        super().__init__(config, model)
+        self.model = model
+        self.config = config
+
+    @staticmethod
+    def build_model(config: Config[MaxOfN]) -> HookedTransformer:
+        simpler_cfg = HookedTransformerConfig(
+            d_model=config.d_model,
+            n_layers=config.n_layers,
+            n_heads=config.n_heads,
+            d_head=config.d_head,
+            n_ctx=config.n_ctx,
+            d_vocab=config.d_vocab,
+            seed=config.seed,
+            attn_only=True,
+            normalization_type=None,
+            # device=default_device(deterministic=config.deterministic),
+        )
+        model = HookedTransformer(simpler_cfg)
+        if config.zero_biases:
+            for name, param in model.named_parameters():
+                if "b_" in name:
+                    param.requires_grad = False
+        return model
+
+    @staticmethod
+    def loss_fn(
+        logits: Float[Tensor, "batch pos d_vocab"],
+        tokens: Integer[Tensor, "batch pos"],
+    ) -> Float[Tensor, ""]:
+        logits = logits[:, -1, :]
+        true_maximum = torch.max(tokens, dim=1)[0]
+        log_probs = logits.log_softmax(-1)
+        correct_log_probs = log_probs.gather(-1, true_maximum.unsqueeze(-1))
+        return -correct_log_probs.mean()
+
+    @staticmethod
+    def acc_fn(
+        logits: Float[Tensor, "batch pos d_vocab"],
+        tokens: Integer[Tensor, "batch pos"],
+    ) -> float:
+        pred_logits = logits[:, -1, :]
+        pred_tokens = torch.argmax(pred_logits, dim=1)
+        true_maximum = torch.max(tokens, dim=1)[0]
+        return (pred_tokens == true_maximum).float().mean().item()
+
+    def run_batch(self, x: Float[Tensor, "batch pos"], prefix: str):
+        self.model.to(x.device, print_details=False)
+        # print(self.model.)
+        # print(x.device)
+        y_preds = self.model(x)
+        loss = self.loss_fn(y_preds, x)
+        self.log(f"{prefix}loss", loss, prog_bar=True)
+        acc = self.acc_fn(y_preds, x)
+        self.log(f"{prefix}acc", acc, prog_bar=True)
+        return loss
+
+    def training_step(self, batch, batch_idx):
+        return self.run_batch(batch, prefix="")
+
+    def test_step(self, batch, batch_idx):
+        self.run_batch(batch, prefix="test_")
+
+    def configure_optimizers(self):
+        return torch.optim.Adam(
+            self.parameters(),
+            lr=self.config.optimizer_kwargs["lr"],
+            betas=self.config.optimizer_kwargs["betas"],
```

Comment:
More compactly generalizable alternative (pick whichever seems better to you):
```suggestion
            **{k:self.config.optimizer_kwargs[k] for k in ("lr", "betas")}
```